### ISSUE TITLE
Returning LiveData from domain layer

### ISSUE THREAD

**Issue by pablobaldez** [android10/Android-CleanArchitecture-Kotlin#14]
What do you think about returning LiveData instances directly from domain Layer?
I'm asking it because the one of the best features of Room framework is the hability to observe changes in DB directly.

So the LiveData instances should be returned directly from domain (usecase and repositories)?

**Comment by St4B** [comment]
LiveData is part of the framework so i would not use it in the domain layer.

**Comment by Zhuinden** [comment]
AFAIK livedata can work even in unit tests?

**Comment by pablobaldez** [comment]
@St4B I don't think is a good aproach lose Room features of update LiveData just because its part of framework.
This rule is very important to make the unit tests easier, but we can test livedata using Junit without any problem, just like another project class.

**Comment by St4B** [comment]
hmmm ... I was affected by the old project (with rxJava) which had different models per layers and mappers. Now that we use the same model in every layer it does not seem wrong to me. Basically I started to like it! : p

**Comment by pablobaldez** [comment]
@Zhuinden yes. You have to add dependencies of livedata to unit test
`android.arch.core:core-testing:$arch_version`
and than add the rule `InstantTaskExecutorRule` into unit test file

**Comment by android10** [comment]
Totally agree here with the above comments. 
 
```LiveData``` belongs to the Android Framework which is something you wanna avoid as much as you can, at least at domain level. 

In my opinion, it clearly belong to the UI layer, being a key part of MVVM.

**Comment by nikolajakshic** [comment]
If you use RxJava in your project, the solution might be to return the Observable from the domain layer and use the reactive streams to convert it to the LiveData in the UI layer.

**Comment by crjacinro** [comment]
> If you use RxJava in your project, the solution might be to return the Observable from the domain layer and use the reactive streams to convert it to the LiveData in the UI layer.

Is this available already in the source code? This would be a big help since I want to take advantage of Room -> LIve Data interaction by using this clean arch repository.

**Comment by SaeedMasoumi** [comment]
I agree that the domain layer must not have a dependency on the platform-specific frameworks/libraries, but based on YAGNI principle we're just adding an extra level of abstraction without gain.  
So I think passing `LiveData` in domain layer give us some advantages like less boilerplate for some features like Room.

**Comment by mitevyav** [comment]
I am kind a new to the Clean Approach, but there was always something that made my crazy in the begging when I tried to understand it. Everywhere, like in here, everybody are saying - just plain Java! No Android, not this, no that, but every time RxJava is the solution of everything. But RxJava is not pure Java!! I swear, I have never seen an article saying - "Here we are using RxJava, and we are breaking the principle, but there are more pros than cons, so it is OK!". So saying this I don't see why RxJava is ok, but LiveData not.

**Comment by nikolajakshic** [comment]
@mitevyav Well, RxJava is not something that is Android specific, it can be used everywhere (web, mobile, desktop, etc..), on the other side, `LiveData` is Android specfic, that's why it's "acceptable" to return something like `Observable` from the domain layer, and for `LiveData` it is not. It's just that RxJava is so popular, widely accepted and some will say it feels almost like a part of the language.

IMO, you don't need to strictly follow clean "rules", just pick what suits your use-case the best and as long as your app is testable and maintainable you are good.

**Comment by lawloretienne** [comment]
> @Zhuinden yes. You have to add dependencies of livedata to unit test
> `android.arch.core:core-testing:$arch_version`
> and than add the rule `InstantTaskExecutorRule` into unit test file

After doing that it throws an error that has do with with Powermockito.

**Comment by lawloretienne** [comment]
So i commented out these dependencies

//    testImplementation “org.mockito:mockito-all:$mockitoVersion”
//    testImplementation “org.hamcrest:hamcrest-all:$hamcrestVersion”
//    testImplementation “org.powermock:powermock-module-junit4:2.0.0-RC.1”
//    testImplementation “org.powermock:powermock-api-mockito:1.7.4"

and now i just have these depenedencies

testImplementation “junit:junit:$junitVersion”
testImplementation “org.mockito:mockito-core:2.21.0"

and there is no longer a conflict.

**Comment by IMDroidude** [comment]
Why don't you go for Flow instead of LiveData. as Flow is a pure data driven item.

**Comment by Zhuinden** [comment]
> LiveData is part of the framework so i would not use it in the domain layer.

Objectively, if you don't want to use LiveData, then you shouldn't use Room.

Room is also Android-specific.

Honestly, we should just all use Flutter, you can port that to any platform now. 😂

**Comment by vasilyev04** [comment]
Definitely, at the moment there is only one solution for this problem and it's using Coroutine Flow instead of LiveData.

That's it!

**Comment by Zhuinden** [comment]
RxJava / Reaktive are technically both valid options

**Comment by vasilyev04** [comment]
> RxJava / Reaktive are technically both valid options

I suppose that it's not necessary to bring huge framework in your project just for gaining data from Room :)

**Comment by Zhuinden** [comment]
In that case, just use LiveData 😉

**Comment by vasilyev04** [comment]
> In that case, just use LiveData 😉

The question is "how to avoid liveData in domain layer?", so..
